flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutorToResourceManagerConnection.java

```diff
@@ -93,22 +94,24 @@ public void start() {
 		registration.startRegistration();
 
 		Future<Tuple2<ResourceManagerGateway, TaskExecutorRegistrationSuccess>> future = registration.getFuture();
+
+		ExecutionContext executionContext = taskExecutor.getRpcService().getExecutionContext();
 		
 		future.onSuccess(new OnSuccess<Tuple2<ResourceManagerGateway, TaskExecutorRegistrationSuccess>>() {
 			@Override
 			public void onSuccess(Tuple2<ResourceManagerGateway, TaskExecutorRegistrationSuccess> result) {
-				registeredResourceManager = result.f0;
 				registrationId = result.f1.getRegistrationId();
+				registeredResourceManager = result.f0;
 			}
-		}, taskExecutor.getMainThreadExecutionContext());
+		}, executionContext);
```

Comment:
If we establish the contract that you first check `registeredResourceManager` before accessing `registrationId` it should be thread-safe wrt the `registeredResourceManager` and `registrationId` fields. We could do that more explicitly by returning `null` in `getRegistrationId` if `registeredResourceManager` is not set.

Additionally, we would have to guard the `start` method which should be doable without extra costs since it will be started only by one thread anyway.